Second source-only build of VMR introduces different reference packages

There is something that is causing different reference packages to be added to the `prereqs/packages/reference` directory when a second source-only build occurs on the VMR. This doesn't happen on the first build, only on subsequent builds.

### Repro Steps

1. Checkout the release/10.0.1xx-preview3 branch of the VMR
2. Make some source change in any repo (except the sdk repo) that will cause a build error.
3. Run `./prep-source-build.sh && ./build.sh --source-only --online`. This will encounter the build error.
4. Undo your source change.
5. Run `./build.sh --source-only --online`

This will cause the following error in the sdk repo:

```
    CSC : error CS0006: Metadata file '/repos/dotnet/src/sdk/artifacts/sb/package-cache/microsoft.aspnetcore.app.ref/8.0.0/analyzers/dotnet/cs/Microsoft.AspNetCore.App.Analyzers.dll' could not be found [/repos/dotnet/src/sdk/src/BuiltInTools/BrowserRefresh/Microsoft.AspNetCore.Watch.BrowserRefresh.csproj]
    CSC : error CS0006: Metadata file '/repos/dotnet/src/sdk/artifacts/sb/package-cache/microsoft.aspnetcore.app.ref/8.0.0/analyzers/dotnet/cs/Microsoft.AspNetCore.App.CodeFixes.dll' could not be found [/repos/dotnet/src/sdk/src/BuiltInTools/BrowserRefresh/Microsoft.AspNetCore.Watch.BrowserRefresh.csproj]
    CSC : error CS0006: Metadata file '/repos/dotnet/src/sdk/artifacts/sb/package-cache/microsoft.aspnetcore.app.ref/8.0.0/analyzers/dotnet/cs/Microsoft.AspNetCore.Components.Analyzers.dll' could not be found [/repos/dotnet/src/sdk/src/BuiltInTools/BrowserRefresh/Microsoft.AspNetCore.Watch.BrowserRefresh.csproj]
    CSC : error CS0006: Metadata file '/repos/dotnet/src/sdk/artifacts/sb/package-cache/microsoft.aspnetcore.app.ref/8.0.0/analyzers/dotnet/roslyn4.4/cs/Microsoft.Extensions.Logging.Generators.dll' could not be found [/repos/dotnet/src/sdk/src/BuiltInTools/BrowserRefresh/Microsoft.AspNetCore.Watch.BrowserRefresh.csproj]
    CSC : error CS0006: Metadata file '/repos/dotnet/src/sdk/artifacts/sb/package-cache/microsoft.aspnetcore.app.ref/8.0.0/analyzers/dotnet/roslyn4.4/cs/Microsoft.Extensions.Options.SourceGeneration.dll' could not be found [/repos/dotnet/src/sdk/src/BuiltInTools/BrowserRefresh/Microsoft.AspNetCore.Watch.BrowserRefresh.csproj]
```

This occurs because it's getting an incorrect version of the Microsoft.AspNetCore.App.Ref.8.0.0 package, one that doesn't contain the fix from https://github.com/dotnet/source-build-reference-packages/pull/1199. Somehow this older version of Microsoft.AspNetCore.App.Ref.8.0.0 is being copied to the `prereqs/packages/reference` directory and picked up in the build. This is likely coming from the previously source built artifacts? It seems like when the build starts the `prereqs/packages/reference` directory is populated with the ref packages from PSB artifacts. Then SBRP builds and those get cleared out with the current ones. When the subsequent build occurs, it clears the current ones out and replaces them with the ones from the PSB artifacts. Then it skips SBRP build since that was already built and you're left with incorrect ref packages.

--- a/src/SourceBuild/content/Directory.Build.props
+++ b/src/SourceBuild/content/Directory.Build.props
@@ -178,6 +178,8 @@
     <ResultingPrebuiltPackagesDir>$([MSBuild]::NormalizeDirectory('$(PackageReportDir)', 'prebuilt-packages'))</ResultingPrebuiltPackagesDir>
     <SbrpRepoSrcDir>$([MSBuild]::NormalizeDirectory('$(SrcDir)', 'source-build-reference-packages', 'src'))</SbrpRepoSrcDir>
     <ReferencePackagesDir>$([MSBuild]::NormalizeDirectory('$(PrereqsPackagesDir)', 'reference'))</ReferencePackagesDir>
+    <PreviouslySourceBuiltReferencePackagesDirName>SourceBuildReferencePackages</PreviouslySourceBuiltReferencePackagesDirName>
+    <PreviouslySourceBuiltReferencePackagesDir>$([MSBuild]::NormalizeDirectory('$(PrebuiltSourceBuiltPackagesPath)', '$(PreviouslySourceBuiltReferencePackagesDirName)'))</PreviouslySourceBuiltReferencePackagesDir>
     <SourceBuiltArtifactsTarballName>Private.SourceBuilt.Artifacts</SourceBuiltArtifactsTarballName>
     <SourceBuiltPrebuiltsTarballName>Private.SourceBuilt.Prebuilts</SourceBuiltPrebuiltsTarballName>
 

--- a/src/SourceBuild/content/eng/finish-source-only.proj
+++ b/src/SourceBuild/content/eng/finish-source-only.proj
@@ -269,7 +269,7 @@
           DestinationFolder="$(SourceBuiltLayoutDir)"
           UseSymbolicLinksIfPossible="true" />
     <Copy SourceFiles="@(ReferencePackageToBundle)"
-          DestinationFolder="$(SourceBuiltLayoutDir)SourceBuildReferencePackages"
+          DestinationFolder="$(SourceBuiltLayoutDir)$(PreviouslySourceBuiltReferencePackagesDirName)"
           UseSymbolicLinksIfPossible="true" />
 
     <!-- Content of the .version file to include in the tarball -->

--- a/src/SourceBuild/content/eng/init-source-only.proj
+++ b/src/SourceBuild/content/eng/init-source-only.proj
@@ -47,32 +47,9 @@
           Condition="'@(SourceBuiltPrebuiltsTarballFile)' != ''" />
   </Target>
 
-  <!-- Copy SBRP packages to reference packages location. -->
-  <Target Name="CopySourceBuiltReferencePackages"
-          DependsOnTargets="UnpackSourceBuiltArtifactsArchive"
-          Inputs="$(PrebuiltSourceBuiltPackagesPath)SourceBuildReferencePackages"
-          Outputs="$(ReferencePackagesDir)">
-    <ItemGroup>
-      <UnpackedSourceBuildReferencePackage Include="$(PrebuiltSourceBuiltPackagesPath)SourceBuildReferencePackages/*" />
-    </ItemGroup>
-
-    <!-- When building iteratively (e.g. making infra changes), the directory may already exist and needs to be cleaned up. -->
-    <RemoveDir Directories="$(ReferencePackagesDir)" />
-
-    <!-- Either move or copy the unpacked SBRP packages.
-         Don't move when the packages directory is externally provided. -->
-    <Move SourceFiles="@(UnpackedSourceBuildReferencePackage)"
-          DestinationFolder="$(ReferencePackagesDir)"
-          Condition="'$(CustomPrebuiltSourceBuiltPackagesPath)' == ''" />
-    <Copy SourceFiles="@(UnpackedSourceBuildReferencePackage)"
-          DestinationFolder="$(ReferencePackagesDir)"
-          Condition="'$(CustomPrebuiltSourceBuiltPackagesPath)' != ''" />
-  </Target>
-
   <Target Name="UnpackTarballs"
           DependsOnTargets="UnpackSourceBuiltArtifactsArchive;
-                            UnpackSourceBuiltPrebuiltsArchive;
-                            CopySourceBuiltReferencePackages" />
+                            UnpackSourceBuiltPrebuiltsArchive" />
 
   <!-- Build the custom msbuild sdk resolver. -->
   <Target Name="BuildMSBuildSdkResolver"
@@ -96,8 +73,8 @@
   <Target Name="ExtractToolsetPackages" DependsOnTargets="UnpackTarballs">
     <ItemGroup>
       <ToolsetPackage Include="Microsoft.DotNet.Arcade.Sdk" SourceFolder="$(PrebuiltSourceBuiltPackagesPath)" Version="$(ARCADE_BOOTSTRAP_VERSION)" />
-      <ToolsetPackage Include="Microsoft.Build.NoTargets" SourceFolder="$(ReferencePackagesDir)" Version="$(NOTARGETS_BOOTSTRAP_VERSION)" />
-      <ToolsetPackage Include="Microsoft.Build.Traversal" SourceFolder="$(ReferencePackagesDir)" Version="$(TRAVERSAL_BOOTSTRAP_VERSION)" />
+      <ToolsetPackage Include="Microsoft.Build.NoTargets" SourceFolder="$(PreviouslySourceBuiltReferencePackagesDir)" Version="$(NOTARGETS_BOOTSTRAP_VERSION)" />
+      <ToolsetPackage Include="Microsoft.Build.Traversal" SourceFolder="$(PreviouslySourceBuiltReferencePackagesDir)" Version="$(TRAVERSAL_BOOTSTRAP_VERSION)" />
     </ItemGroup>
 
     <Unzip SourceFiles="%(ToolsetPackage.SourceFolder)%(ToolsetPackage.Identity).%(ToolsetPackage.Version).nupkg"

--- a/src/SourceBuild/content/eng/tools/Directory.Build.props
+++ b/src/SourceBuild/content/eng/tools/Directory.Build.props
@@ -5,7 +5,7 @@
   <!-- Don't use these feeds when testing. Projects under this directory are referenced by
        test projects. This makes sure that the feeds aren't used. -->
   <PropertyGroup>
-    <RestoreSources Condition="'$(DotNetBuildSourceOnly)' == 'true' and '$(Test)' != 'true'">$(ReferencePackagesDir);$(PrebuiltPackagesPath);$(PrebuiltSourceBuiltPackagesPath)</RestoreSources>
+    <RestoreSources Condition="'$(DotNetBuildSourceOnly)' == 'true' and '$(Test)' != 'true'">$(PreviouslySourceBuiltReferencePackagesDir);$(PrebuiltPackagesPath);$(PrebuiltSourceBuiltPackagesPath)</RestoreSources>
   </PropertyGroup>
 
 </Project>

